public/static/plugins/perf/js/trend_chart.js

```diff
@@ -141,31 +161,44 @@ var drawSingleTrendChart = function(params) {
       height: cfg.container.height,
     })
 
+  var colors = d3.scale.category10();
+  // FIXME Force color range 'initialization'. Might be a bug in d3 3.5.3
+  // For some reason, d3 will not give you, let's say, the third color
+  // unless you requested 0, 1 and 2 before.
+  for (var i = 0; i < cfg.knownLevelsCount; i++) colors(i);
+
   var ops = _.pluck(series, cfg.valueAttr);
   var opsValues = _.pluck(series, 'ops_per_sec_values');
   var avgOpsPerSec = d3.mean(ops)
 
   var allLevels = _.pluck(series[0].threadResults, 'threadLevel')
-  var levels = threadMode == MAXONLY
-    ? [_.max(allLevels, function(d) { return +d })]
-    : allLevels
-
-  var levelsMeta = _.map(levels, function(d, i) {
-    var data = {
-      name: d,
-      idx: _.indexOf(allLevels, d),
-    };
-
-    var match = cfg.knownLevels[d];
-
-    data.colorId = match
-      ? match.colorId
-      : cfg.knownLevelsCount + i
-    return data
-  })
+
+  var levelsMeta = threadMode == MAXONLY
+    ? [{name: 'MAX', idx: cfg.valueAttr, color: '#484848'}]
+    : _.map(allLevels, function(d, i) {
+      var data = {
+        name: d,
+        idx: _.indexOf(allLevels, d),
+      };
+
+      var match = cfg.knownLevels[d];
+
+      data.color = colors(match ? match.colorId : cfg.knownLevelsCount + i)
+      return data
+    })
 
   // Calculate legend x pos based on levels
-  cfg.legend.xPos = (cfg.container.width - levels.length * cfg.legend.step) / 2
+  cfg.legend.xPos = (cfg.container.width - levelsMeta.length * cfg.legend.step) / 2
+
+  // Returns value extractor fn for given `level` and series `item`
+  // This function is curried, so you should call it as fn(level)(item)
+  var getValueFor = (function() {
+    if (threadMode == MAXONLY) {
+      return PerfChartService.getValueForMaxOnly
+    } else {
+      return PerfChartService.getValueForAllLevels
+    }
+  })()
```

Comment:
Hi Ruslan. I'm going to abuse this opportunity to learn something about JavaScript. Why is the above better than the simpler form:

var getValueFor;
if (threadMode == MAXONLY) {
  getValueFor = PerfChartService.getValueForMaxOnly
} else {
  getValueFor = PerfChartService.getValueForAllLevels
}

Reply:
@henrikingo, I moved this code to several places and tried different approaches, and just forget to simplify it when made a final changes. In this context, there is no reason to use closures. And your code is almost equivalent and a bit simpler, but I would use ternary if expression.